src/osd/PG.cc

```diff
@@ -5700,12 +5700,21 @@ bool PG::can_discard_replica_op(OpRequestRef& op)
   const T *m = static_cast<const T *>(op->get_req());
   assert(m->get_type() == MSGTYPE);
 
+  int from = m->get_source().num();
+
+  // if a repop is replied after a replica goes down in a new osdmap, and
+  // before the pg advances to this new osdmap, the repop replies before this
+  // repop can be discarded by that replica OSD, because the primary resets the
+  // connection to it when handling the new osdmap marking it down, and also
+  // resets the messenger sesssion when the replica reconnects. to avoid the
+  // out-of-order replies, the messages from that replica should be discarded.
```

Comment:
Can you explain where the out-of-order reply comes from?  AFAICS it should be fine to pretend we are still living in the past since the PG hasn't advanced yet.  If the peer reconnects, it will get rejected... and if the peer restarts, any message it sends will be marked with a later epoch and we won't process it until we advance maps.